### PR TITLE
emsg_multiline: log Vim errors

### PR DESCRIPTION
**Pull request by blueyed**
Might provide more insight in case of errors (e.g. https://github.com/neovim/neovim/issues/10182#issuecomment-521326982).

### CODE REVIEW

**Comment by justinmk** [comment]
> Might provide more insight in case of errors (e.g. [#10182 (comment)](https://github.com/neovim/neovim/issues/10182#issuecomment-521326982)).

CI only logs ERROR log-level. Any higher level is too noisy for CI (and causes builds to be much slower, specially on macOS).

**Comment by blueyed** [comment]
I see.
Is it slower mostly due to explicit flushing?!

Anyway, I've thought it should not be logged at level error by default, should it?

An alternative might also be to explicitly use "error"/ELOG on CI.

**Comment by justinmk** [comment]
> Is it slower mostly due to explicit flushing?!

Yes. But even if not, it would make the logs too noisy to dump at the end of the CI build log.

> Anyway, I've thought it should not be logged at level error by default, should it?

Just try it, and see if it shows up in the CI log :)

**Comment by blueyed** [comment]
> > Anyway, I've thought it should not be logged at level error by default, should it?
> 
> Just try it, and see if it shows up in the CI log :)

?
What I've meant that we should not use `ELOG` here, since it is not a real (fatal) error after, but could just be `:echoerr "foo"` etc.
It might be OK though?

**Comment by justinmk** [comment]
I know. But if this is to debug https://github.com/neovim/neovim/issues/10182#issuecomment-521326982 , then it can be temporary.

The PR in general doesn't really make sense to me, it seems to have overlap with 'verbose' option (aka `nvim -V[1-0] log`).

**Comment by blueyed** [comment]
> I know. But if this is to debug [#10182 (comment)](https://github.com/neovim/neovim/issues/10182#issuecomment-521326982) , then it can be temporary.

Yeah, ok.  Although that appears to happen very rarely only.  I.e. it would be temporary for longer.

The functionality overlaps with `verbosefile` - it's just around the patch (IIRC).
I've found it very interesting / useful, but mostly since there's a log (for me) always, but I am not using verbosefile usually.  And certainly errors messages appear to be as useful as jobs being started therein.  It provides context around other log messages(, and e.g. "echoerr 'stamp'" can be used to add marks etc.)

File: src/nvim/message.c [hunk 1]
```diff
@@ -582,9 +582,24 @@ static bool emsg_multiline(const char *s, bool multiline)
...
+      if (sourcing_name != NULL && sourcing_lnum != 0) {
+        DLOG("%s (%s (line %ld))", s, sourcing_name, (long)sourcing_lnum);
+      } else {
+        DLOG("%s", s);
+      }
+
       return true;
     }
 
+    // Log errors as warnings.
```
Comment (justinmk) [inline src/nvim/message.c]:
```suggestion
    // Log editor errors as INFO.
```

File: src/nvim/message.c [hunk 2]
```diff
@@ -582,9 +582,24 @@ static bool emsg_multiline(const char *s, bool multiline)
         }
         redir_write(s, strlen(s));
       }
+
+      // Log (silent) errors as debug messages.
+      if (sourcing_name != NULL && sourcing_lnum != 0) {
+        DLOG("%s (%s (line %ld))", s, sourcing_name, (long)sourcing_lnum);
+      } else {
+        DLOG("%s", s);
```
Comment (justinmk) [inline src/nvim/message.c]: these should have some sort of indication that they were silenced.
```suggestion
        DLOG("(:silent) %s", s);
```

**Comment by justinmk** [comment]
That's true. I've thought about "unifying" 'verbose' and NVIM_LOG_FILE somehow. 

This is a reasonable step for now, let's do it.

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
What about this additionally?

To log deadly signals:

> preserve_exit:973: Vim: Caught deadly signal 'SIGTERM'

```diff
diff --git i/src/nvim/misc1.c w/src/nvim/misc1.c
index 44e2c7df5..9a3f4c109 100644
--- i/src/nvim/misc1.c
+++ w/src/nvim/misc1.c
@@ -970,6 +970,7 @@ void preserve_exit(void)
   mch_errmsg(IObuff);
   mch_errmsg("\n");
   ui_flush();
+  ILOG("%s", IObuff);

   ml_close_notmod();                // close all not-modified buffers
```

OTOH it might be problematic due to memory allocation etc, and therefore should be avoided probably.

But Neovim should/could log the signal in/with "mch_exit:99: Nvim exit: 1".

Another issue though.

**Pull request merged**